### PR TITLE
Component Generation

### PR DESCRIPTION
**Pull request by plt-joey**
Create a separate project publishable as a `dotnet tool` for component generation.

Update 2021-07-01: Added samples for generation via string manipulation and generation via AST

Update 2021-08-20: Code complete and ready for first-pass reviews

NOTES TO REVIEWERS:
- I made minor changes to Dash.NET itself that the generated components require to function. If you are testing a generated component in a dash app you have to point it at the version of Dash.NET in this PR, and not the one on nuget
- For that same reason I have not added anything to the build scripts for publishing this tool yet, the updated version of Dash.NET has to be published first
- Using the generated components also requires a minor change to the dash app template ([!2](https://github.com/plotly/Dash.NET.Template/pull/2)), `contentRoot` needs to point to the build output, and not to the source code:
```fsharp
let contentRoot = Reflection.Assembly.GetExecutingAssembly().Location |> Path.GetDirectoryName
```

Rough todo list, in no particular order
- [x] Create new project based on existing `BackboneGenerator.fsx`
- [x] Refactor existing generator
    - [x] Switch to AST based generation
    - [x] Add property generation
    - [x] Add property Enum and Union(?) type handling
    - [x] Generate documentation
    - [x] Switch components to Feliz-style instantiation (see [!18](https://github.com/plotly/Dash.NET/issues/18))
    - [x] Harden
- [x] Add:
    - [X] Automatic component `.js` script loading on import/use
    - [X] Project generation 
    - [X] Build pipeline generation 
    - [x] Nuget package definition generation
    - [x] CLI
- [x] Parse information from `metadata.json`
    - [x] Add support for flow types and ts types
- [x] Add package/publish steps for the `dotnet tool`
- [x] Automatic component publishing? (may want to separate this into a second task/PR)
- [x] Unit Tests
- [x] Fix Build Scripts
- [x] Documentation

### CODE REVIEW

**Comment by plt-joey** [comment]
There is an issue with the unit test when run in `build-and-test-linux`
Namely the two errors:

`'WebClient' is required here and is unavailable. You must add a reference to assembly 'System.Net.WebClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'.`

and 

`typecheck error The type 'RequireQualifiedAccess' is not defined.`

These don't seem to affect the windows unit tests and I am unable to replicate them locally.

The offending tests use F# pre-compilation operations to check if the generated code is valid, and I believe these errors are just a mismatch in environment.

Edit: this has been fixed

**Comment by plt-joey** [comment]
Discovered a couple bugs while running component generation on dash core components:

- Complex type dependencies would sometimes be defined in the wrong order
- The `Option<List<Something>>` type would fail to generate correctly
- Certain characters would break automatic documentation ('<','>', and '`')
- DU case names cannot contain a '+', even if escaped

These have been fixed.

**Comment by plt-joey** [comment]
Changed type names and case names to convert to PascalCase from snake_case instead of just capitalizing the first letter, for consistency.

**Comment by plt-joey** [comment]
Improved the naming scheme of types to make them easier to use out of the box with no edits to the generated code, as well as added more helper feliz-style constructors for union types

**Comment by plt-joey** [comment]
I seem to have misunderstood what `ObjectOf` was. Doing more testing to fix

**Comment by plt-joey** [comment]
> I seem to have misunderstood what `ObjectOf` was. Doing more testing to fix

`ObjectOf` has been changed to a single case union of a dictionary, and now works correctly

Note: The constructor I am using for the `Dictionary` doesn't seem to exist in eyes of the unit tests, but it is in the [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.-ctor?view=net-5.0#System_Collections_Generic_Dictionary_2__ctor_System_Collections_Generic_IEnumerable_System_Collections_Generic_KeyValuePair__0__1___) and compiles fine outside of the unit test. Fixing this unit test is a TODO

Edit: This unit test issue was fixed when I switched from `Dictionary` to `Map`

**Comment by plt-joey** [comment]
Misunderstood how the json conversion worked, switched the uses of `ToString` to instead use a `Convert` method that creates a version of the property that is nicely jsonifyable

**Comment by plt-joey** [comment]
Handful of bugs/inconsistencies discovered during core component conversion that need to be fixed before merging

Bugs:
- `&` in XML documentation should be converted to `&amp;`
- `option` values are not converted to json correctly
- All convert functions should box the object values

Inconsistencies:
- Generated code should use `Map` not `Dictionary`
- `.Convert()` should be a static `|> Type.convert`
- `init` and `applyMembers` don't use the type-safe types
- The arguments of `init` and `applyMembers` are not converted to camelCase

Edit (Aug 31, 2021): These have been addressed

**Comment by kunjee17** [comment]
Looks good to me.

File: src/Dash.NET/DashApp.fs [hunk 1]
```diff
@@ -112,15 +113,63 @@ type DashApp =
     ///
     /// POST /_dash-update-component -> handles callback requests and returns serialized callback JSON responses.
     static member toHttpHandler (app:DashApp) : HttpHandler =
+
+        //TODO: put this component loading somewhere nicer when the Giraffe setup is abstracted
```
Comment (plt-joey) [inline src/Dash.NET/DashApp.fs]: now that #20 has been merged in this can be moved to a more sensible location

Comment (plt-joey) [reply src/Dash.NET/DashApp.fs]: fixed

**Comment by plt-joey** [comment]
Rebase broke the generator unit tests with the error `typecheck error The type 'DashHtmlEngine' does not define the field, constructor or member 'Html'.`

edit: fixed

File: src/Dash.NET.ComponentGeneration/DocumentationGeneration.fs [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+    | ArrayOf (_, Some value) -> 
+        value
+        |> getTypePropDocumentation
+        |> Option.map (sprintf "list with values of type: %s")
+
+    | ObjectOf (_, Some value) -> 
+        value
+        |> getTypePropDocumentation
+        |> Option.map (sprintf "dict with values of type: %s")
```
Comment (alexcjohnson) [inline src/Dash.NET.ComponentGeneration/DocumentationGeneration.fs:83]: dict -> record?

(apologies for the late comment, just poking around here to see how you implemented things)

Comment (plt-joey) [reply src/Dash.NET.ComponentGeneration/DocumentationGeneration.fs:83]: not in this case. `ObjectOf` is actually a `Map`, which is the F# version of `IDictionary`. I think I put `dict` here to be consistent with the python generator, but this technically should be `Map`.

Comment (alexcjohnson) [reply src/Dash.NET.ComponentGeneration/DocumentationGeneration.fs:83]: Got it - just wanted to make sure this wasn't an unintended remnant from Python :)

**Pull request merged**